Objetivo 7

Balrrach opened the pull request:
* [x] ¿Se han registrado todos los eventos relevantes en la aplicación, es decir, se han añadido comandos para registrarlos en todos ellos?
* [x] ¿Se han seguido buenas prácticas en el uso de configuración remota?
* [x] ¿Se han explicado correctamente las decisiones técnicas que se han tomado sobre las bibliotecas usadas?
* [x] ¿Te has asegurado de que el fichero `.env` no se pueda añadir accidentalente al repositorio?
* [x] ¿Los logs y la configuración están en un objeto que se *inyecta* en las clases manejadoras para su uso?
* [x] ¿*Realmente* se usan los logs en las diferentes clases en eventos importantes?
* [x] ¿Se testea el código creado para logs y configuración, igual que cualquier otro código?
* [x] ¿La clase de configuración lee *siempre* las variables de entorno?

Balrrach committed:
Clean

Es conveniente mantener las discusiones de todas las
herramientas en un mismo fichero

Balrrach committed:
.

Balrrach committed:
Feat #30

Es necesario justificar la elección de la librería de logging
para poder revisarla en un futuro y porque la justificación
fomenta la búsque da una librería que se adecúe a los requisitos
del sistema.

Balrrach committed:
Fix

Se pretende que los tests solo corran cuando terminen las GHA
encargadas de crear el contenedor o de actualizarlo.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Feat #30

Se ha reconsiderado la elección puesto que ulog no tiene ssoporte
nativo para typescript.

Balrrach committed:
Clean

Balrrach committed:
Feat #30

Ni los logs ni los archivos de configuración pueden terminar
en el repositorio.

Balrrach committed:
Feat #30

Las decisiones técnicas relacionadas con las librerías
utilizadas para la configuración remota y para la carga de las
variables de entorno han de estar justificadas para su futura
revisión.

Balrrach committed:
Clean

Balrrach committed:
Clean

Balrrach committed:
Clean

Balrrach committed:
.

Balrrach committed:
Clean

El mensaje es más claro

Balrrach committed:
Feat #30

El orden en el que se toman las configuraciones es:
1. Servidor
2. Enviroment
3. Valores predeterminados
De esta forma la aplicación siempre tiene una configuración
preparada para utilizar.

Puesto que no se puede devolver una promesa de un objeto dada
su propiedad inherente de *disponibilidad(Readiness)*, la
solución que se ha tomado es crear una atributo que corresponde
a una promesa y ejecutar los metodos necesarios cuando la
promesa se verifique. La solución detallada se puede encontrar
en: https://stackoverflow.com/questions/35743426/async-constructor-functions-in-typescript

Balrrach committed:
Clean

Tiene más sentido tener la ruta en primer lugar puesto que conforma la
primera parte de la ruta.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Clean

Balrrach committed:
Fix

No se están utilizando los mismos tags en run_test que en
create_pr_container puesto que no se están usando los mismos
eventos(on: pull request ===vs=== on: push)

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Feat #30, #33

Es conveniente utilizar una clase controladora e instanciarla
al inicio del programa asegurándose de esta manerar que todas sus
dependencias están correctamente inicializadas desde el primer momento.

Balrrach committed:
Feat #30, #33

La clase controladora se exporta para su uso en el resto de clases.
Se implementa getLogger() puesto que es necesario acceder al logger.

Balrrach committed:
Clean

Balrrach committed:
Feat #30, #33

Primer intento de implentacion de un controlador asincrono.
No funciona correctamente porque, por un motivo desconocido, no se está
esperando a que la instancia config del controlador se termine de
instanciar antes de utilizarla en el index o en los test aunque el test
específico de la clase configuración funciona cocorrectamente.

Balrrach committed:
.

Balrrach committed:
Fix

No se puede mantener la estructura de un contenedor por PR porque JJ
verifica sus tests contra el contenedor con tag latest

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

JJ commented:
⛹ Revisores → @joaquingv12 @alexespana @Asmilex

Balrrach commented:
Revisión @JJ.

Balrrach commented:
Tengo un problema con la asincronía de la clase controlador. No consigo que espere a la ejecución de la creación correcta de la clase config. Si alguno ve el fallo le agradecería que me lo comentara.

Balrrach committed:
.

Balrrach committed:
Feat #33

La dependencia no es del método sino de estado de disponibilidad del
objeto.

Balrrach committed:
Clean

No es neceasario esperar a que el objeto de configuración esté preparado
en index puesto que no se hace un uso explicito del controlador.

Balrrach committed:
Closes #34

El problema era consecuencia de que el código que encerraba la promesa
de la clase Config tiraba las llamadas asíncronas e inmediatamente
resolvía a undefined. Como consecuencia el estado Ready no resolvía el
problema. La solución ha sido considerar Ready como el conjunto de
promesas con las que realmente está ligado. Más información en:
https://stackoverflow.com/questions/70501382/related-promises-in-object-creation

Balrrach committed:
Feat #34

De misma forma ocurre con el Controlador. El logger no tenía por qué
estar instanciado lo estaba por el mero hecho de que el objeto de
configuración se instanciaba automáticamente por error. Al resolver su
asincronía el esto ya no tiene que ser así y hay que asegurarse que el
logger esta correcamente instanciado.

No termino de entender porque estaba corriendo correctamente antes de
este arreglo.

Balrrach committed:
.

Asmilex commented on docs/tools.md:
```diff
@@ -0,0 +1,92 @@
+Uno de los sistemas que permite dicho uso es [etcd](https://etcd.io/). Y una de las librerías que inplementa su API(más bien etcd3 que es una interfaz de alto nivel que permite la interación con etcd) es [etcd3](https://www.npmjs.com/package/etcd3).
+La librería está implementada por Microsoft lo cual es positivo puesto que lo usa en sus propios proyectos y es más difícil que se abandone a largo plazo.
+Además, tiene muy pocos issues y PR abiertos y todos los PR están relacionados con la actualización de dependencias mientras que ningún issue está relacionado con bugs sino con enriquecimientos y otros.
+Queda fuera de las miras de la asignatura la configuración de un servidor pero se ha simulado su funcionamiento en la aplicación. Para acerlo funcionar correctamente bastaría con instanciar el objeto con los valores de un puerto de un servidor correctamente configurado.
```
*"acerlo"*. Oops.

Balrrach replied:
Arreglado en df3ffc6.

Asmilex commented on docs/tools.md:
```diff
@@ -0,0 +1,92 @@
+
+
+## Variables de Entorno
+Es necesario cargar la configuración a partir de un fichero local o de la propia terminal. Para esto se puede parsear un archivo/entrada estandard o se puede utilizar una librería que realice esta función.
```
*"estandard"* -> *estándar*.

Balrrach replied:
Arreglado en df3ffc6.

Asmilex commented on docs/tools.md:
```diff
@@ -0,0 +1,92 @@
+El primero es que el estado de mantenimiento de dotenv es mucho mejor que el de nconf; aunque la última versión disponible de dotenv es algo más antigua que la de nconf este último tiene la friolera de un 98 issues abiertos a 26 de diciembre de 2021. Además, cuanto más se avanza en el historial de issues se observa que estos no son respondidos por ninguno de los mantenedores.
+El segundo es que, mientras que dotenv no tiene dependencias, nconf tiene 4. Cuantas menos dependencias la librería mejor porque menos dependencias tendrá el proyecto.
+El tercero es que dotenv tiene el soporte para tipos totalmente integrado mientras que nconfg las tiene en un paquete a parte(esto es un detalle).
+
```
Creo que está muy bien documentado, la verdad. Rápido, conciso y directo al grano 👍🏼.

Asmilex commented on iv.yaml:
```diff
@@ -5,3 +5,4 @@ automatizar:
   orden: npm run
 test: src/tests/location.test.ts
 CI: .github/workflows
+configuracion: config/configuration.env
```
Aquí tiene que ir el archivo de Typescript que se encarga de gestionar la configuración, no el archivo de variables de entorno. De hecho, este tipo de archivos no se deben cargar en el repo. 

Tuve el mismo fallo yo en su momento, jeje.

Balrrach replied:
Cierto, me he equivocado. Arreglado en d3406ab.

Asmilex commented on src/classes/route.ts:
```diff
@@ -9,6 +13,10 @@ class Route {
 		this.route = route;
 		this.orders = orders;
+
+		loggerPromise.then(loggerReady => {
+			loggerReady.info("Route object correctly instantiated");
+		})
```
La forma de invocar al logger parece un poco tosca. No controlo la librería, pero en general, el uso de loggers no debería requerir tirar de promesas constantemente. La sintaxis deberían ser algo del estilo

```
logger.nivel("Blablabla)"
```

Quizás sea por la limitación de TS, no lo sé. Pero es algo que quizás merezca la pena darle una vuelta de más.

Balrrach replied:
Es consecuencia de permitir que la configuración se haga de forma asíncrona. El logger lo da la configuración que, al ser asíncrona, responde con una promesa. A partir de este punto es necesario trabajar con promesas, no veo otra alternativa.

JJ replied:
Tienes que hacer `await` de esa promesa para recibir el "payload"

Asmilex commented on src/classes/route.ts:
```diff
@@ -1,5 +1,9 @@
 import { Location } from './location';
 import { Order } from './order'
+import { controller } from '../controller'
+const loggerPromise = controller.Ready.then(() => {
+	return controller.getLogger().child({ module: 'Route'})
+})
```
Hay un par de detalles que me gustaría comentar con respecto a la forma de invocar los loggers:

## Sobre la instanciación

Según tengo entendido, la forma de organizar el logger es la siguiente:

```
Controlador crea un logger usando la configuración -> se reparte mediante inyección de dependencias en el resto de clases
```

Aquí, por lo que veo, se utiliza esta estructura. Sin embargo, me da la sensación de que se instancia un nuevo logger hijo cada vez. Es decir, que por cada archivo, se está creando un logger derivado del *maestro* que reside en el controlador. 

Quizás es porque no comprenda bien del todo cómo funciona la librería y/o TS, pero quizás merezca la pena echarle un ojillo. 

## Inyección de dependencias y principio open/closed 

Lo que sí está claro es que el logger reside en el módulo. Esto, en principio, no es demasiado correcto. ¿Es esta la forma indicada de hacerlo en TS? 

En mi caso, tuvimos una discusión por hacerlo de manera similar en Python. 

La solución a la que llegamos nosotros fue crear una clase (estilo pipeline, workflow, etc) que gestione las clases individuales. Según el principio open/closed, no se deberían modificar las clases que creaste anteriormente en el sentido de añadir dependencias. Se debe crear una nueva capa de abstracción que gestione el logging. 

En principio, esto afecta también al resto de archivos.

Balrrach replied:
> ## Sobre la instanciación
> 
> Según tengo entendido, la forma de organizar el logger es la siguiente:
> 
> ```
> Controlador crea un logger usando la configuración -> se reparte mediante inyección de dependencias en el resto de clases
> ```
> 
> Aquí, por lo que veo, se utiliza esta estructura. Sin embargo, me da la sensación de que se instancia un nuevo logger hijo cada vez. Es decir, que por cada archivo, se está creando un logger derivado del _maestro_ que reside en el controlador.
> 
> Quizás es porque no comprenda bien del todo cómo funciona la librería y/o TS, pero quizás merezca la pena echarle un ojillo.

Lo entiendes bien, se instancia un logger hijo parar cada clase porque permite incluír el contexto de la clase en todos los logs que genere el hijo. No veo el problema.

> ## Inyección de dependencias y principio open/closed
> 
> Lo que sí está claro es que el logger reside en el módulo. Esto, en principio, no es demasiado correcto. ¿Es esta la forma indicada de hacerlo en TS?
> 
> En mi caso, tuvimos una discusión por hacerlo de manera similar en Python.
>

¿Por qué no es correcto?

Asmilex replied:
> Lo entiendes bien, se instancia un logger hijo parar cada clase porque permite incluír el contexto de la clase en todos los logs que genere el hijo. No veo el problema.

Yo ahí veo mucho overhead innecesario. Eso hace que tengas un logger por cada archivo heredado del padre. Cuando tengas 30 archivos, tienes 30 loggers diferentes, y todos ellos luchando por escribir en los mismos sitios. Uf, dunno, pero lo veo chungo. 

Creo que lo correcto es tener un mismo logger que se pasa entre los diferentes módulos. De esa forma, no tienes 30 promesas de que te llegará un logger hijo. Que la clase controladora instancie el logger necesario y lo reparta entre los que lo pidan. Espera a que llegue la promesa, y cuando esté, la devuelves. De esa forma tienes únicamente uno. 

> ¿Por qué no es correcto?

Pues, sinceramente, no te sé dar una respuesta contundente. La misma duda tuve yo en Python, y mis correctores me sugirieron quitar el logger del módulo.

Balrrach replied:
> Yo ahí veo mucho overhead innecesario.
¿Overhead instanciar 30 objetos? No creo que sea para tanto. Se puede hacer la prueba con cronómetro para salir de dudas.

> Eso hace que tengas un logger por cada archivo heredado del padre. Cuando tengas 30 archivos, tienes 30 loggers diferentes, y todos ellos luchando por escribir en los mismos sitios. Uf, dunno, pero lo veo chungo.
Si la asincronía está bien implementada no debería haber problema. La escritura de los loggers va a seguir siendo secuencial da igual que haya 1 que 1000 porque la promesa es la misma para todos. En cualquier caso sólo hay un acceso simultáneo al archivo porque el programa no está paralelizado y, aunque lo estuviera, sería una cuestión del sistema operativo que no debe fallar.

> Creo que lo correcto es tener un mismo logger que se pasa entre los diferentes módulos. De esa forma, no tienes 30 promesas de que te llegará un logger hijo. Que la clase controladora instancie el logger necesario y lo reparta entre los que lo pidan. Espera a que llegue la promesa, y cuando esté, la devuelves. De esa forma tienes únicamente uno. 
Ensonces no tiene sentido hacer el controlador asíncrono puesto que lo primero que se va a hacer después de instanciar al controlador es instanciar los loggers y el programa quedaría automaticamente parado. Puede que el problema sea que realmente no tiene sentido montar todo esto por tener un controlador asíncrono en una única llamada a un servidor y/o en una única llamada I/O.

> Pues, sinceramente, no te sé dar una respuesta contundente. La misma duda tuve yo en Python, y mis correctores me sugirieron quitar el logger del módulo.
A ver si @joaquingv12, @alexespana o @JJ arrojan algo de luz.

Balrrach replied:
He decidido modificar la estructura para el logger no resida en el módulo. Ahora sigue un patrón sigletón y cada clase pide la instancia al comienzo del fichero. De esta forma se tiene certeza de que hay un único controlador en el programa en ejecución.

Asmilex reviewed (commented):
A excepción de un par de decisiones que tenemos que mirar, creo que está todo muy bien. Especialmente la documentación.

Balrrach committed:
Fix: Typo

Balrrach committed:
Fix

El campo configuración tiene que hacer referencia al fichero que maneja
la configuración no al fichero que la contiene(que ni siquiera se sube
al repositorio).

Balrrach committed:
Fix: Typo

Balrrach committed:
Clean

Lo mejor es que todas las variables tenga el tipo más cerrado posible
para aprovechar al máximo las características del tipado fuerte.

Balrrach committed:
Feat #30, #33

Se ha decidido transformar la creación del configurador en síncrona
porque en caso contrario todas las llamadas al mismo(en esencia todo el
programa porque es necesario para instanciar el logger, el servidor...)
se convertían asíncronas haciendo su gestión terriblemente tediosa para
las supuestas ventajas que podía tener la asincronía.

JJ commented:
Tus propios tests están fallando...

Balrrach commented:
El único motivo por el que los test no pasan es porque el programa no tiene permiso para crear el directorio temporal en el que almacenar los logs.
Mañana arreglo el Dockerfile.

Balrrach committed:
Feat #30, #33

Es conveniente añadir un archivo de configuracion de prueba al proyecto
para correr los test de la clase de configuración y para que los desarrolladores
tengan una guía para la creación de su propio archivo de configuracón.

Balrrach committed:
Feat #30, #33

Es conveniente añadir un archivo de configuracion de prueba al proyecto
para correr los test y para que los desarrolladores tengan una guí

Balrrach committed:
Merge branch 'Objetivo-7' of github.com:Balrrach/IV-Proyecto into Objetivo-7

Balrrach committed:
Fix

Se arregla el gitignore para incluir el archivo de configuración de
prueba.

Balrrach commented:
Ya está solucionado...

JJ commented on config/configuration_test.env:
```diff
@@ -0,0 +1,2 @@
+LOG_DIR = '../tmp/logs/'
```
No tienes garantizado que tengas derecho de escritura ahí. Los valores por omisión deberían de ser un poco más generales.

Balrrach replied:
Tienes razón. El problema es que por defecto no se puede escribir en el repositorio al que le hace checkout la github action. Aunque he estado investigando, creo que no es posible dar permisos de escritura sobre el repositorio porque no es clonado. En cuanto encuentre una solución cambiaré la ruta. He creado #39 para el seguimiento de este problema.

JJ commented on src/classes/command.ts:
```diff
@@ -1,5 +1,9 @@
 import { Restaurant } from './restaurant'
+import { Controller } from '../controller'
+
+const logger = (Controller.getInstance()).getLogger().child({ module: 'Command'})
```
Las variables globales de módulo nunca son una buena idea.

Balrrach replied:
Tienes razón. Convertirlo en un atributo de la clase no me parece una idea mejor porque el problema real es que es una dependencia transversal y, por lo tanto, si cambiara cualquier cosa que tenga que ver con esta(como la forma de loggear) habría que arreglar una cantidad ingente de código. La conclusión a la que he llegado es que lo mejor es encapsularlo de la mejor manera y en typescript creo que una buena forma es por medio de decoradores. El problema con esta solución es que la interfaz tiene que permitir todas las operaciones que se quieran hacer con el logger y todavía esto no lo tengo totalmente claro. He creado #38 para el seguimiento de este problema.

JJ commented on src/classes/delivery_man.ts:
```diff
@@ -1,9 +1,12 @@
 import { Route } from './route';
+import { Controller } from '../controller'
+
+const logger = (Controller.getInstance()).getLogger().child({ module: 'DeliveryMan'})
```
Y esta forma de acceder es excesivamente complicada, y la estás repitiendo siempre. Deberías tener un API que te permitiera instanciarlo sólo con el nombre de la clase (y esto debería ser también automático, por cierto)

Balrrach replied:
Mirar #38.

JJ commented on src/classes/destination.ts:
```diff
@@ -15,6 +18,8 @@ class Destination extends Location {
 		this.letter = letter;
 		this.customerName = customerName;
+
+		logger.info("Destination object correctly instantiated");
```
Estás modificando todas y cada una de las clases para logear, en vez de hacerlo desde los módulos que creen o destruyen objetos de esas clases. Por el principio open/closed, no deberías. hacer esto.

Balrrach replied:
Tienes razón, los métodos públicos no deberían loggearse ellos mismos.

JJ commented on src/controller.ts:
```diff
@@ -0,0 +1,49 @@
+import { Config } from "./config";
+import { pino } from 'pino';
+import * as fs from 'fs';
+import { connect } from 'http2';
```
Esto no se usa.

Balrrach replied:
Se me pasó eliminarlo.

JJ reviewed (approved):
Los objetivos principales se han conseguido, pero realmente como integras el logger en la aplicación tendrás que cambiarlo para el siguiente objetivo. Las clases no pueden preocuparse de loguear lo que les ocurre, sino quien las cree, por *separation of concerns* y también el principio open/closed que hemos repetido tantas veces.

Balrrach commented:
Todos los problemas para los que tenga una respuesta clara serán resueltos inmediatamente el PR del Objetivo-8. Todos lo problemas para los que no tengo una respuesta clara tienen su correspondiente issue.

Balrrach commented:
Gracias @Asmilex por tus revisiones.
